### DISCUSSION TITLE
How can I use a pre-render approach with a Lua document filter?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#9646] [Q&A] [answered]
### Description

I'm thinking about using the [pre-render](https://quarto.org/docs/projects/scripts.html#pre-and-post-render) of Quarto to modify a document to create a [tabset panel](https://quarto.org/docs/output-formats/html-basics.html#tabsets) that includes how the code cell was rendered. My goal for doing this is to avoid copying and pasting cell contents to show how the cell was made. I suppose it's an extended/alternative cell-escape sequence that retains the computational order.

For instance, take a Quarto document like:

**Original**

````qmd
---
title: "My Code Cell Demo"
format: html
---

```{python}
#| show-cell: true
print("Hello, world!")
```

````

With **Pre-render** script:

````qmd
---
title: "My Code Cell Demo"
format: html
---

:::{.panel-tabset}

#### Rendered

```{python}
#| show-cell: true
print("Hello, world!")
```

#### Source

```{{python}}
#| show-cell: true
print("Hello, world!")
```

:::

````

I'd like to know if there is a document-facing Lua example that receives the AST for the document. I see there is a note on using the [embedded Deno](https://quarto.org/docs/projects/scripts.html#deno-scripts) that I would like to avoid using.

**Comment by mcanouil** [comment]
Pre-render happens before the document is translated into the AST I believe.

This being said, pre-render supports several languages in which you can parse and edit text.

PS: be aware that Pandoc has parsing rules requesting headings and code blocks to be surrounded by empty lines.

**Reply by coatless** [reply]
Okay, don't mind the parsing portions at the moment. I'll clean that up.

A few more questions then: 

1. With the `pre-render` stage does this mean I'm modifying the _source_/_original_ document?
    - Is it not a copy that Quarto is storing in memory or under a temporary folder? 
2. Is there a different stage of the render process that I should be targeting? 
    - The focus is being able to obtain the code cell _before_ it is executed. 
3. Regarding the Lua transcription, I'm seeing a Deno example of: 

```js
const config = Deno.readTextFileSync("_quarto.yml"); /* removed parse() from YML */
```

I suppose the equivalent would be reading with Pandoc's [`io`](https://www.lua.org/manual/5.3/manual.html#6.8) into a text file and, then, applying some regex to identify the line in the cell? (Brittle, I'm aware).

**Reply by mcanouil** [reply]
1. If you don't want to modify the source files, then you need some holistic to copy/restore the files first. Note that 1.5, scan files a second time after pre-render, which means you can create entirely new documents at pre-render.
2. Pre-render is the only thing that happens before computation (some shortcodes are also evaluated before computations such as the include shortcode).
3. I am not very familiar with Deno and more generally TypeScript, but yes that's a the example in <https://quarto.org/docs/projects/scripts.html#deno-scripts>.

PS: I assume `echo: fenced` does not meet your needs here, but mentioning it just in case: <https://quarto.org/docs/computations/execution-options.html#fenced-echo>.

**Reply by coatless** [reply]
Mmm, is there a changelog entry on the scan files portion or documentation page I can read through? That sounds very useful for what I'm interested in. 

Unfortunately, [`echo: fenced`](https://github.com/quarto-dev/quarto-web/blob/5de4536c369c9d73dad0388eb97b3e61c40754de/docs/computations/execution-options.qmd#L284-L294) option doesn't go as far as I need since the code cell is not being replicated. Though, it's a great alternative (that I wasn't expecting!) to using double curly escape clauses, e.g. 

````md
```{{<language>}}
```
````

**Reply by mcanouil** [reply]
I was thinking about the echo option, as the result could be targeted by a LUA filter.
For instance, if the cell contains the option and "show-cell", copy the code block and insert tabset.

> Mmm, is there a changelog entry on the scan files portion or documentation page I can read through? That sounds very useful for what I'm interested in.

It's in the change log of 1.5 and there is a small page linked right at the top of the changelog under "highlights": <https://quarto.org/docs/prerelease/1.5/pre-render.html>.

**Reply by coatless** [reply]
Mmm, thinking more... The scanning approach is useful for shifting the `.qmd`'s into a temporary directory. Though, I'm not sure copying the entire directory would be ideal. 

Going back over to the `echo: fenced`, this probably would work well for an extension targeting just built-in languages; but, I'd also like to use it for custom cells `<lang1>-<extended>`. I might still throw together a quick extension built on that idea. 

Though, for a cell marked with the `-`, that would knock off the computational order. Is there a way to inject at the point where a Lua filter modifies the AST a new compute region?

**Reply by mcanouil** [reply]
> Though, for a cell marked with the -, that would knock off the computational order. Is there a way to inject at the point where a Lua filter modifies the AST a new compute region?

I don't think so. (cc. @cscheid)

**Reply by cscheid** [reply]
I'm not sure I understand the question.

**Reply by coatless** [reply]
@cscheid 

1. How can code cells be inserted into the document _before_ evaluation happens? 
   - Evaluation refers to the cells being run in the order with which the cells can be found in the document.
2. If code cells need to be inserted into the document at a later time, is there a way to re-trigger the render? (Similar to the [`quarto.ConditionalBlock` API](https://quarto.org/docs/prerelease/1.3/custom-ast-nodes/conditional-block.html).)

**Reply by cscheid** [reply]
There isn't a great mechanism to do that at the moment, because all Lua filters run after the engine executes.

**Reply by coatless** [reply]
If I switch over to Deno/Typescript, same issue?

**Reply by cscheid** [reply, marked as answer]
With that said, here's a Lua filter that does the specific thing you want (to rearrange a cell output into tabset panels). 

It turns out that Tabsets can be constructed directly in Lua code with `quarto.Tabset`.

There's a small Quarto bug that needs fixing, but the filter below includes a workaround. Hopefully the filter is self-explanatory.

````
---
format: html
keep-md: true
filters:
  - topanel.lua
---


::: {.to-panel}

```{r}
#| echo: true

print("Hello, world")
```

:::
````

### topanel.lua

```lua
function Div(div)
  if not div.classes:includes("to-panel") then 
    return
  end
  local code_block = nil
  local cell_output = nil
  quarto._quarto.ast.walk(div.content, {
    CodeBlock = function(code)
      if code.classes:includes("cell-code") then
        code_block = code
      end
    end,
    Div = function(div)
      if div.classes:includes("cell-output") then
        cell_output = div
      end
    end
  })

  local rendered = quarto.Tab({ title = "Rendered", content = cell_output })
  local source = quarto.Tab({ title = "Source", code_block })
  local tabs = pandoc.List({ rendered, source })

  return quarto.Tabset({
    level = 3,
    tabs = tabs,
    attr = pandoc.Attr("", {"panel-tabset"}) -- this shouldn't be necessary but it's the bug I mentioned.
  })
end
```

**Reply by cscheid** [reply]
> If I switch over to Deno/Typescript, same issue?

Yes. There really isn't an extension point before Lua filters. With that said, your specific problem has a more specific solution, see above.

**Reply by coatless** [reply]
@cscheid thank you so much! If you ever are in the bay, let me know I owe you a coffee at the very least! 

One last question (sorry!), are the `quarto.Tab`/`quarto._quarto.ast.walk` documented somewhere?

**Reply by cscheid** [reply]
(Paraphrasing)

> is `quarto.Tab` documented somewhere?

This should be, it's an oversight. We do document `quarto.Tabset` in the Lua typing intelligence we provide, but both `quarto.Tabset` and `quarto.Tab` should be documented in general.

> is `quarto._quarto.ast.walk` documented somewhere?

This isn't, and I shouldn't have used it whoops :) You don't need to use that in this case, `div:walk ({...})` will work. The difference in our `ast.walk` method is that finds and processes custom AST nodes (I'm just used to it from developing internal code). We haven't documented its pitfalls just yet, but we also should.
